<a href="https://colab.research.google.com/github/Yfyangd/BMIR/blob/master/HW2_Optional.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Import Library

In [0]:
from tensorflow import keras
import tensorflow as tf
from tensorflow.keras import datasets, layers, models

# Import Fashion Minst Datasets

In [2]:
fashion_mnist = keras.datasets.fashion_mnist
(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()
class_names = ['T-shirt/top', 'Trouser', 'Pullover', 'Dress', 'Coat','Sandal', 'Shirt', 'Sneaker', 'Bag', 'Ankle boot']
train_images = train_images.astype('float32') / 255
train_images = train_images.reshape(train_images.shape[0], 28, 28, 1)
test_images = test_images.astype('float32') / 255
test_images = test_images.reshape(test_images.shape[0], 28, 28, 1)
train_labels = train_labels.reshape(train_labels.shape[0], 1)
test_labels = test_labels.reshape(test_labels.shape[0], 1)

4423680/4422102 [==============================] - 0s 0us/step


In [3]:
train_images.shape

(60000, 28, 28, 1)

# Traditional neural network of HW1
Test accuracy is **0.8827**

In [13]:
model = keras.Sequential([
    keras.layers.Flatten(input_shape=(28, 28)),
    keras.layers.Dense(128, activation='relu'),
    keras.layers.Dense(10)
])

model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

model.fit(train_images, train_labels, epochs=10)

test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)

print('\nTest accuracy:', test_acc)

Epoch 1/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.4977 - accuracy: 0.8251
Epoch 2/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3745 - accuracy: 0.8652
Epoch 3/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3390 - accuracy: 0.8762
Epoch 4/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.3123 - accuracy: 0.8852
Epoch 5/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2949 - accuracy: 0.8916
Epoch 6/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2805 - accuracy: 0.8957
Epoch 7/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2694 - accuracy: 0.8995
Epoch 8/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2585 - accuracy: 0.9033
Epoch 9/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.2488 - accuracy: 0.9079
Epoch 10/10
1875/1875 [==============================] - 3s 2ms/step - loss: 0.239

# Model: LeNet-5

In [0]:
model = keras.Sequential()
model.add(layers.Conv2D(filters=6, kernel_size=(3, 3), activation='relu', input_shape=(28,28,1)))
model.add(layers.AveragePooling2D())
model.add(layers.Conv2D(filters=16, kernel_size=(3, 3), activation='relu'))
model.add(layers.AveragePooling2D())
model.add(layers.Flatten())
model.add(layers.Dense(units=128, activation='relu'))
model.add(layers.Dense(units=64, activation='relu'))
model.add(layers.Dense(units=10, activation = 'softmax'))

# Model Training

In [22]:
model.compile(optimizer='adam',
              loss=tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True),
              metrics=['accuracy'])

history = model.fit(train_images, train_labels, epochs=10, 
                    validation_data=(test_images, test_labels))

Epoch 1/10
1875/1875 [==============================] - 5s 3ms/step - loss: 1.7215 - accuracy: 0.7470 - val_loss: 1.6786 - val_accuracy: 0.7853
Epoch 2/10
1875/1875 [==============================] - 5s 3ms/step - loss: 1.6519 - accuracy: 0.8101 - val_loss: 1.6706 - val_accuracy: 0.7903
Epoch 3/10
1875/1875 [==============================] - 5s 3ms/step - loss: 1.6358 - accuracy: 0.8259 - val_loss: 1.6429 - val_accuracy: 0.8195
Epoch 4/10
1875/1875 [==============================] - 5s 3ms/step - loss: 1.6236 - accuracy: 0.8378 - val_loss: 1.6391 - val_accuracy: 0.8212
Epoch 5/10
1875/1875 [==============================] - 5s 3ms/step - loss: 1.6117 - accuracy: 0.8497 - val_loss: 1.6243 - val_accuracy: 0.8369
Epoch 6/10
1875/1875 [==============================] - 5s 3ms/step - loss: 1.6030 - accuracy: 0.8584 - val_loss: 1.6087 - val_accuracy: 0.8534
Epoch 7/10
1875/1875 [==============================] - 5s 3ms/step - loss: 1.5953 - accuracy: 0.8663 - val_loss: 1.6045 - val_accuracy:

In [23]:
test_loss, test_acc = model.evaluate(test_images,  test_labels, verbose=2)
print("Part_3_fashion_mist: test accuracy: ", test_acc)

313/313 - 1s - loss: 1.5881 - accuracy: 0.8724
Part_3_fashion_mist: test accuracy:  0.8723999857902527


# Result
* Compare to the traditional neural network of HW1 (accuracy: **0.8791**), the test accuracy of LeNet-5 is **0.8724**.
* 理論上Convolution Network performance應該要比傳統的Neural Network要高, 但是在這邊LeNet-5 準確度 與 HW-1傳統model 相差不多的原因可能如下:
> * LeNet-5 convolution 只有2層, 特徵提取不太足夠. 在我的Homework2中, 使用3層convolution 的準確度為0.9074, 高於LetNet-5 與 傳統的Nerual network
> * 對於解析度比較低的圖片, 傳統的Neural network全部平坦化的節點數不多, 沒有convolution提取特徵的缺點比較不顯著
> * LeNet-5 使用的是Average Pooling, 在homework-2使用的是Max Pooling, 這可能也是影響performance的原因